### PR TITLE
Make incomings payloads public

### PR DESCRIPTION
**Pull request by pcreux**
Hello friends!

In order to process messages asynchronously, I want to store their `payload` in a queue. In this PR, I make the `@payload` instance variable public.

It however conflicts with the existing `Postback#payload` method.

How would you like to go about that?

1. Rename `@payload` to `@raw_payload` ?
2. Rename `Postback#payload` to `Postback#fb_payload`?
3. ... ?

### CODE REVIEW

File: spec/facebook/messenger/incoming/postback_spec.rb [hunk 1]
```diff
@@ -18,6 +18,12 @@
 
   subject { Facebook::Messenger::Incoming::Postback.new(payload) }
 
+  describe '.payload' do
```
Comment (pcreux) [inline spec/facebook/messenger/incoming/postback_spec.rb]: 💥 
```
  1) Facebook::Messenger::Incoming::Postback.payload returns the original payload
     Failure/Error: expect(subject.payload).to eq(payload)

       expected: {"sender"=>{"id"=>"3"}, "recipient"=>{"id"=>"3"}, "timestamp"=>1457764197627, "postback"=>{"payload"=>"USER_DEFINED_PAYLOAD"}}
            got: "USER_DEFINED_PAYLOAD"

       (compared using ==)

       Diff:
       @@ -1,5 +1,2 @@
       -"postback" => {"payload"=>"USER_DEFINED_PAYLOAD"},
       -"recipient" => {"id"=>"3"},
       -"sender" => {"id"=>"3"},
       -"timestamp" => 1457764197627,
       +"USER_DEFINED_PAYLOAD"
     # ./spec/facebook/messenger/incoming/postback_spec.rb:23:in `block (3 levels) in <top (required)>'
```

**Comment by dthyresson** [comment]
Perhaps a variant of 

> 2. Rename Postback#payload to Postback#fb_payload?

Such as: `callback` to match the `Postback Callback` terminology used in FB in https://developers.facebook.com/docs/messenger-platform/webhook-reference#postback

`Postback.callback` would contain what is currently the `payload'?

**Comment by jgorset** [comment]
@t0ggah was actually working on this exact thing at the office today and ran into the same problem. We discussed either using `to_hash` or `raw` or `raw_payload`. What do you guys think?

**Comment by pcreux** [comment]
`raw_payload`?

**Comment by jgorset** [comment]
I think Facebook refers to all incoming requests as "callbacks", though, @dthyresson. It makes sense semantically, but since there's no `callback` attribute in the JSON from Facebook it's less intuitive than `postback`.

**Comment by jgorset** [comment]
Does `raw_payload` imply that it's unprocessed? If so, should it return a JSON string? If so, should to `to_hash` return it parsed into a hash? Am I overthinking this? Or asking too many questions? What is the meaning of life, anyway?

**Comment by pcreux** [comment]
Right... Well... I would go with `raw_payload` and `42`.

**Comment by jgorset** [comment]
Haha, alright. I'm cool with `raw_payload`!

**Comment by pcreux** [comment]
Ok, so I've renamed `payload` to `raw_payload` everywhere and it's quite heavy loaded... What about renaming `Postback#payload` to `Postback#postback_payload`? (It's heavy loaded... but just in postback...)

**Comment by dthyresson** [comment]
So :) I'll throw another idea out there in the "naming is hard" ... for arguments sake.

If we'd like to match

```
I think Facebook refers to all incoming requests as "callbacks", though, @dthyresson. It makes sense semantically, but since there's no callback attribute in the JSON from Facebook it's perhaps less intuitive than payload.
```
the JSON attributes, when I look at:

```
{
   "object":"page",
   "entry":[
      {
         "id":PAGE_ID,
         "time":1458668856451,
         "messaging":[
            {
               "sender":{
                  "id":USER_ID
               },
               "recipient":{
                  "id":PAGE_ID
               },
               "delivery":{
                  "mids":[
                     "mid.1458668856218:ed81099e15d3f4f233"
                  ],
                  "watermark":1458668856253,
                  "seq":37
               }
            }
         ]
      }
   ]
}
```

or 

```
{
  "object":"page",
  "entry":[
    {
      "id":PAGE_ID,
      "time":1458692752478,
      "messaging":[
        {
          "sender":{
            "id":USER_ID
          },
          "recipient":{
            "id":PAGE_ID
          },
          "timestamp":1458692752478,
          "postback":{
            "payload":"USER_DEFINED_PAYLOAD"
          }
        }
      ]
    }
  ]
}
```

what's being "received" isn't a payload, but the `messaging` key.

As in https://github.com/hyperoslo/facebook-messenger/blob/master/lib/facebook/messenger/server.rb#L40:

```
      def receive
        return if app_secret && !integrity?

        hash = JSON.parse(@request.body.read)

        # Facebook may batch several items in the 'entry' array during
        # periods of high load.
        hash['entry'].each do |entry|
          # Facebook may batch several items in the 'messaging' array during
          # periods of high load.
          entry['messaging'].each do |messaging|
            Facebook::Messenger::Bot.receive(messaging)
          end
        end
      end
```

Is that what's being passed in to `incoming/postback` etc ... the `message/messaging` vs the `payload`?

Since the  Bot receives "messaging" 

```
Facebook::Messenger::Bot.receive(messaging)
```

And it's there in https://github.com/hyperoslo/facebook-messenger/blob/master/lib/facebook/messenger/bot.rb#L46

it becomes a "payload" per the argument name .. which is then made into a callback :)

```
        def receive(payload) 
          callback = Facebook::Messenger::Incoming.parse(payload)

          case callback
          when Incoming::Message then trigger(:message, callback)
          when Incoming::Delivery then trigger(:delivery, callback)
          when Incoming::Postback then trigger(:postback, callback)
          when Incoming::Optin then trigger(:optin, callback)
          end
        end
```

Maybe `message` vs `payload`?

That way the nested attribute of `payload` in the Postback can remain `payload` ... *and* the attributes match the incoming FB JSON?

**Comment by pcreux** [comment]
@dthyresson Following your reasoning, I would suggest using `messaging` instead of `payload` in `Incoming::*` classes.

**Comment by dthyresson** [comment]
@pcreux Yep, that's a shorter way of saying it 😄

**Comment by pcreux** [comment]
Ok, friends ,`messaging` it is. Feel free to squash that little PR.

**Comment by jgorset** [comment]
I like it; `messaging` is exactly in line with Facebook's terms, and since we're such a thin wrapper I think we should always strive for that. Nice job!

**Comment by jgorset** [comment]
:sparkles: :cake: :sparkles:

**Pull request merged**